### ISSUE TITLE
zip behavior

### ISSUE THREAD

**Issue by olsonpm** [dtao/lazy.js#135]
Is the following behavior intended in zip?  I googled for a tiny bit and there didn't seem to be any official implementation.

```
Lazy([1,2,3]).zip([1,2]) // [[1, 1], [2, 2], [3, undefined]
Lazy([1,2]).zip([1,2,3]) // [[1, 1], [2, 2]
```

**Comment by djcsdy** [comment]
FWIW this differs from the implementation of `zip` in both Haskell and .NET (Linq), both of which truncate the result to the length of the shortest input.

**Comment by olsonpm** [comment]
If any maintainers want this changed, I'll create a PR for a fix.

**Comment by bchociej** [comment]
The no-truncation behavior matches that of Underscore's zip, which is probably intended giving Lazy's heredity

**Comment by olsonpm** [comment]
@bchociej - I just tested it on jsfiddle and I assume you skimmed my example.  Underscore doesn't truncate in either scenarios whereas lazy.js truncates if the first array is shorter than the second array.  I didn't test more than two arrays.

**Comment by bchociej** [comment]
@olsonpm Sorry, I guess I said that really oddly. I agree with you -- I think that your first example, which matches Underscore, is probably the intent. In which case, I agree that the second example should give an array of length 3! I meant my comment to be in contrast to @djcsdy 's examples where the output is of the minimum length, truncating longer inputs.

**Comment by olsonpm** [comment]
Ah gotcha.  Not a problem - and I think the no-truncation behavior would be most valuable since it allows you to easily compare the equality of two sequences without first knowing the length.

**Comment by dtao** [comment]
Hm, thanks for bringing this up. I think I don't have a *strong* opinion as to which behavior is correct, although to me this actually feels like a scenario where it's likely a mistake on the developer's part. So maybe I *do* have an opinion, which would be *not* to truncate, so as to expose as much information as possible (and also match Underscore's behavior).

Yeah... I think I will update this to follow Underscore. Will close this issue when I've pushed the change.

**Comment by djcsdy** [comment]
Haskell’s behaviour is motivated by type theory (there is no `null` or `undefined` in Haskell), and also by use cases that involve zipping an infinite list with a finite list.

For example, consider the following code to filter out every other item in a list:

```js
Lazy(["Sam", "Sarah", "Susan", "Steve"])
    .zip(Lazy.range(1, Number.POSITIVE_INFINITY))
    .filter(function (pair) { return pair[1] % 2 == 0; })
    .map(function (pair) { return pair[0]; });
// -> Lazy(["Sarah", "Steve"])
```

This construction is quite clumsy with lazy.js, which is a shame, because it can be quite useful.

In Haskell:

```hs
map snd $
    filter (even . fst) $
    zip [1..] ["Sam", "Sarah", "Susan", "Steve"]
-- -> ["Sarah", "Steve"]
```

Neither consideration applies to Underscore, but the above use case does apply to lazy.js and can be quite useful, so it’s worth considering carefully.

**Comment by olsonpm** [comment]
That's interesting.  Thanks much for taking the time to share as I never would have thought of that.

**Comment by jwhitley** [comment]
Folks following this thread who are interested in Haskell style zip semantics may want to check out my change to zip at jacksonfish/lazy.js@f556830, based off of the work in PR #178.

**Comment by dtao** [comment]
This was fixed in version 0.4.3.

**Comment by dtao** [comment]
To clarify: by "fixed" I mean Lazy now behaves the same as Underscore and Lo-dash re: truncation. Generally speaking Lazy isn't super flexible with regards to passing sequences as arguments to methods (sometimes it works, sometimes it doesn't). I intend to revisit this and make the library more consistent on this front in the future; at that time I may take another look at the behavior here and ensure it makes sense when handling things like infinite sequences.
